File: clients/android/app/src/main/java/app/renzoshiori/client/ui/util/AdultFilter.kt
package app.renzoshiori.client.ui.util

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.runtime.State
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue

/**
 * Adult (18+) view filter — the native twin of the web's
 * lib/utils/adult-filter.ts. It filters Library/Browse rendering only; the
 * server-side `nsfwVisibility` setting is a different thing entirely (that one
 * filters the Sources list).
 *
 * The flag is snapshot state, not just a preference read: the web broadcasts a
 * `renzo-hide-adult-changed` event so every mounted view re-filters the moment
 * the menu item is clicked. Reading a plain SharedPreferences boolean inside a
 * composable gives no such signal — the grids would keep showing the old set
 * until the app was restarted.
 */
object AdultFilter {
    private const val PREFS = "renzo_prefs"
    private const val KEY = "renzo_hide_adult"

    /** Tag set the web classifies on, verbatim. */
    // Kept in step with the server's AdultContentClassifier.cs and the web's
    // adult-filter.ts. Explicit RATING/act tags only — ecchi/mature/seinen/josei/
    // harem are deliberately absent (this targets 18+, not fanservice), as are
    // orientation genres and formats.
    private val ADULT_TAGS = setOf(
        "hentai", "erotica", "erotic", "adult", "smut", "pornographic", "porn",
        "18+", "r18", "r-18", "r18+", "r-18g", "nsfw", "adult (18+)", "explicit",
        // Explicit acts/kinks: a source tagging "Blowjob" but not "Adult" would
        // otherwise pass straight through the filter.
        "blowjob", "double penetration", "sex toys", "sexual violence",
        "sexual abuse", "rape", "incest", "netorare", "ntr", "bdsm",
        "bestiality", "futanari", "shemale", "dickgirl", "milf", "nudity",
        // Sexualised minors — never reachable with 18+ hidden.
        "loli", "lolicon", "shota", "shotacon",
    )

    /**
     * The comparable pieces of a raw tag. Sources dress the same rating up in
     * different ways: MangaDex-style prefixes ("Content rating: Pornographic")
     * and bundled alternatives ("Futanari | Shemale | Dickgirl").
     */
    private fun tagParts(raw: String): List<String> {
        var tag = raw.trim()
        val colon = tag.indexOf(':')
        if (colon > 0 && colon < tag.length - 1) {
            val prefix = tag.substring(0, colon).trim().lowercase()
            if (prefix == "content rating" || prefix == "rating" || prefix == "genre") {
                tag = tag.substring(colon + 1).trim()
            }
        }
        val parts = mutableListOf(tag)
        if (tag.contains('|') || tag.contains('/')) {
            tag.split('|', '/').forEach { p -> p.trim().takeIf { it.isNotEmpty() }?.let(parts::add) }
        }
        return parts
    }

    /** True when a single tag name is an explicit 18+ rating. */
    fun isAdultTag(tag: String): Boolean =
        tagParts(tag).any { it.lowercase() in ADULT_TAGS }

    private var hiddenState by mutableStateOf(false)

    /**
     * Loads the persisted flag. Called once from RenzoApp.onCreate.
     *
     * Defaults to HIDDEN. A fresh install shows nothing adult until the user
     * asks for it — the catalogue a source returns is not under our control,
     * so opt-in is the only defensible default (and it's what the store
     * questionnaire is really asking about).
     */
    fun init(context: Context) {
        hiddenState = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getBoolean(KEY, true)
    }

    /** Observable — a composable that reads this recomposes when it changes. */
    fun isHidden(context: Context): Boolean = hiddenState

    fun setHidden(context: Context, hidden: Boolean) {
        hiddenState = hidden
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit().putBoolean(KEY, hidden).apply()
    }

    /** Web `isAdultItem`: prefer the server's flag, fall back to the tags. */
    fun isAdultItem(isNsfw: Boolean?, genres: List<String>): Boolean =
        // OR, not elvis: the web is `isNsfw === true || tags`. With `?:` a FALSE
        // flag short-circuits and the tags are never checked, so where the server
        // flag and the tags disagree this showed what the web hides.
        isNsfw == true || genres.any { isAdultTag(it) }
}

/** Compose-friendly handle mirroring the web's `useHideAdult()` hook. */
class HideAdultState(private val context: Context) {
    /** Derived from the shared flag, so it tracks whoever last changed it. */
    val hidden: State<Boolean> = derivedStateOf { AdultFilter.isHidden(context) }

    fun toggle() = AdultFilter.setHidden(context, !AdultFilter.isHidden(context))
}

@Composable
fun rememberHideAdult(context: Context): HideAdultState = remember { HideAdultState(context) }

File: RenzoFrontend/src/lib/utils/adult-filter.ts
"use client";

import { useSyncExternalStore, useCallback } from "react";

/**
 * Temporary "hide adult content" toggle + genre-based classifier.
 *
 * Classification is tag/genre based and deliberately narrow: explicit adult
 * ratings only (hentai, erotica, smut, …). Ecchi/mature/suggestive are NOT
 * treated as adult — the feature targets 18+ content, not fanservice.
 *
 * The toggle is a client-side view filter, not a setting: state lives in
 * localStorage so "hidden" safely survives reloads, and flipping it back is
 * one click in the user menu. Nothing is deleted or changed server-side.
 */

// Explicit RATING/act tags only. Ecchi, mature, suggestive, seinen, josei and
// harem are deliberately absent — this targets 18+, not fanservice — as are
// orientation genres (yaoi/yuri/BL) and formats (doujinshi), which say what a
// work is about rather than how explicit it is.
//
// Kept in step with the server's AdultContentClassifier.cs and Renzo Hub's
// AdultFilter.kt. All three change together.
const ADULT_TAGS = new Set([
  // Ratings
  "hentai", "erotica", "erotic", "adult", "smut", "pornographic", "porn",
  "18+", "r18", "r-18", "r18+", "r-18g", "nsfw", "adult (18+)", "explicit",
  // Explicit acts/kinks — a source that tags "Blowjob" but not "Adult" would
  // otherwise pass straight through a hide-18+ filter.
  "blowjob", "double penetration", "sex toys", "sexual violence",
  "sexual abuse", "rape", "incest", "netorare", "ntr", "bdsm",
  "bestiality", "futanari", "shemale", "dickgirl", "milf", "nudity",
  // Sexualised minors — never reachable with 18+ hidden.
  "loli", "lolicon", "shota", "shotacon",
]);

/**
 * The comparable pieces of a raw tag. Sources dress the same rating up in
 * different ways and an exact-match set misses all of them: MangaDex-style
 * prefixes ("Content rating: Pornographic") and bundled alternatives
 * ("Futanari | Shemale | Dickgirl", "Fellatio/Blowjob"). Strip the prefix,
 * split the alternatives, and match each piece on its own.
 */
function tagParts(raw: string): string[] {
  let tag = raw.trim();
  const colon = tag.indexOf(":");
  if (colon > 0 && colon < tag.length - 1) {
    const prefix = tag.slice(0, colon).trim().toLowerCase();
    if (prefix === "content rating" || prefix === "rating" || prefix === "genre") {
      tag = tag.slice(colon + 1).trim();
    }
  }
  const parts = [tag];
  if (tag.includes("|") || tag.includes("/")) {
    for (const p of tag.split(/[|/]/)) {
      const t = p.trim();
      if (t) parts.push(t);
    }
  }
  return parts;
}

/** True when the series' genres/tags mark it as explicit adult (18+) content. */
export function isAdultSeries(genres?: string[] | null): boolean {
  if (!genres || genres.length === 0) return false;
  return genres.some((g) => isAdultTag(g));
}

/** True when a single tag name is an explicit adult rating. */
export function isAdultTag(tag: string): boolean {
  if (!tag) return false;
  return tagParts(tag).some((p) => ADULT_TAGS.has(p.toLowerCase()));
}

/**
 * Detection for series/catalog items: prefers the server-computed flag —
 * which aggregates tags across ALL of the item's sources plus the user's
 * manual 18+ override, catching content whose visible source ships no adult
 * tags — and falls back to the visible tags for older cached payloads.
 */
export function isAdultItem(item: { genre?: string[] | null; isNsfw?: boolean }): boolean {
  return item.isNsfw === true || isAdultSeries(item.genre);
}

const STORAGE_KEY = "renzo_hide_adult";
const CHANGE_EVENT = "renzo-hide-adult-changed";

/**
 * Defaults to HIDDEN. A fresh browser (or a fresh install of the native
 * shells, which mirror this key) shows nothing adult until the user asks for
 * it — the catalogue a source returns isn't under our control, so opt-in is
 * the only defensible default. Only an explicit "0" means "show".
 */
function getHideAdult(): boolean {
  if (typeof window === "undefined") return true;
  return localStorage.getItem(STORAGE_KEY) !== "0";
}

function setHideAdult(value: boolean): void {
  if (typeof window === "undefined") return;
  localStorage.setItem(STORAGE_KEY, value ? "1" : "0");
  window.dispatchEvent(new Event(CHANGE_EVENT));
}

function subscribe(callback: () => void): () => void {
  window.addEventListener(CHANGE_EVENT, callback);
  // Cross-tab sync: localStorage writes in other tabs fire `storage`.
  window.addEventListener("storage", callback);
  return () => {
    window.removeEventListener(CHANGE_EVENT, callback);
    window.removeEventListener("storage", callback);
  };
}

/**
 * Reactive hook for the hide-adult toggle. All subscribed components update
 * immediately when any of them (or another tab) flips it.
 */
export function useHideAdult(): [boolean, () => void] {
  const hidden = useSyncExternalStore(subscribe, getHideAdult, () => false);
  const toggle = useCallback(() => setHideAdult(!getHideAdult()), []);
  return [hidden, toggle];
}
